### PR TITLE
Eigenvalue Decay regularization

### PR DESCRIPTION
**Pull request by oswaldoludwig**
I included a new regularizer named Eigenvalue Decay to the deep learning practitioner that aims at maximum-margin learning. This version approximates the dominant eigenvalue by a soft function given by the power method. For details, see:
Oswaldo Ludwig. "Deep learning with Eigenvalue Decay regularizer." ArXiv eprint arXiv:1604.06985 [cs.LG], (2016). https://www.researchgate.net/publication/301648136_Deep_Learning_with_Eigenvalue_Decay_Regularizer

The syntax for Eigenvalue Decay is similar to the other Keras weight regularizers, e.g.:

 model.add(Dense(100, W_regularizer=EigenvalueRegularizer(0.0005)))

### CODE REVIEW

File: keras/regularizers.py [hunk 1]
```diff
@@ -44,7 +67,6 @@ def get_config(self):
                 'l1': self.l1,
                 'l2': self.l2}
 
-
```
Comment (fchollet) [inline keras/regularizers.py]: These lines should not be removed.

File: keras/regularizers.py [hunk 2]
```diff
@@ -15,6 +16,28 @@ def __call__(self, loss):
...
+        self.p = p
+
+    def __call__(self, loss):
+        W = self.p
+        WW = T.dot(W.T,W)
+        dim1, dim2 = WW.shape.eval() # The number of neurons in the layer
+        k = self.k
+        o = np.ones(dim1) # initial values for the dominant eigenvector
+
+        # POWER METHOD FOR APPROXIMATING THE DOMINANT EIGENVECTOR (9 ITERATIONS):
```
Comment (fchollet) [inline keras/regularizers.py]: No need for ALL CAPS comments.

File: keras/regularizers.py [hunk 3]
```diff
@@ -15,6 +16,28 @@ def __call__(self, loss):
...
+        domineigvec = T.dot(WW,T.dot(WW,T.dot(WW,T.dot(WW,T.dot(WW,T.dot(WW,T.dot(WW,T.dot(WW,T.dot(WW,o)))))))))
```
Comment (fchollet) [inline keras/regularizers.py]: PEP8: should have a space after every comma.

File: keras/regularizers.py [hunk 4]
```diff
@@ -15,6 +16,28 @@ def __call__(self, loss):
...
+
+        WWd = T.dot(WW,domineigvec)
```
Comment (fchollet) [inline keras/regularizers.py]: I don't think `domineigvec` is a readable variable name. You should change it to `main_eigenvect` or something similar.

File: keras/regularizers.py (hunk 3 above)
Comment (fchollet) [inline keras/regularizers.py]: Also this line should be broken down into several lines.

File: examples/CIFAR10_with_Eigenvalue_Decay.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
+from __future__ import print_function
```
Comment (fchollet) [inline examples/CIFAR10_with_Eigenvalue_Decay.py]: I don't think this example is necessary. It only slightly differs from the existing CIFAR10 example.

File: keras/regularizers.py (hunk 1 above)
Comment (oswaldoludwig) [inline keras/regularizers.py]: The lines were restored.

File: keras/regularizers.py (hunk 2 above)
Comment (oswaldoludwig) [inline keras/regularizers.py]: All caps comments were rewritten.

File: keras/regularizers.py (hunk 3 above)
Comment (oswaldoludwig) [inline keras/regularizers.py]: The spaces were introduced and the equation of the dominant eigenvector was replaced by a for loop, where it is possible to define the number of iterations of the power method (the GPU/CPU time was evaluated and there is no difference due to this changing).

File: keras/regularizers.py (hunk 4 above)
Comment (oswaldoludwig) [inline keras/regularizers.py]: The variables were renamed as _domin_eigenvect_ and _domin_eigenval_.

File: examples/CIFAR10_with_Eigenvalue_Decay.py (hunk 5 above)
Comment (oswaldoludwig) [inline examples/CIFAR10_with_Eigenvalue_Decay.py]: I agree, the user only need to know examples of the syntax:

from keras.regularizers import EigenvalueRegularizer
...
model.add(Dense(512, W_regularizer=EigenvalueRegularizer(0.1)))

File: keras/regularizers.py [hunk 6]
```diff
@@ -16,6 +18,33 @@ def get_config(self):
...
+class EigenvalueRegularizer(Regularizer):
+    def __init__(self, k):
+        self.k = k
+        self.uses_learning_phase = True
+
+    def set_param(self, p):
+        self.p = p
+
+    def __call__(self, loss):
+        power = 9 # number of iterations of the power method
```
Comment (fchollet) [inline keras/regularizers.py]: PEP8: there should be two spaces before in-line comments

File: keras/regularizers.py [hunk 7]
```diff
@@ -16,6 +18,33 @@ def get_config(self):
         return {'name': self.__class__.__name__}
 
 
+class EigenvalueRegularizer(Regularizer):
```
Comment (fchollet) [inline keras/regularizers.py]: Please add a docstring (for autogenerated documentation)

File: examples/CIFAR10_with_Eigenvalue_Decay.py (hunk 5 above)
Comment (fchollet) [inline examples/CIFAR10_with_Eigenvalue_Decay.py]: Right. Please remove the example.

**Comment by fchollet** [comment]
Also, can you add a quick test for this new regularizer, in `tests/keras/test_regularizers.py`?

File: keras/regularizers.py (hunk 7 above)
Comment (oswaldoludwig) [inline keras/regularizers.py]: A docstring was included in the class EigenvalueRegularizer. I also added a test for the new regularizer in _test_regularizers.py_, here is the output, running on Theano backend in my machine:
======================== test session starts ==================================
platform linux2 -- Python 2.7.6, pytest-2.9.1, py-1.4.31, pluggy-0.3.1
rootdir: /home/oswaldo/Documents/EigenvalueDecay/Eigenvalue_new_version_Keras, inifile: 
collected 3 items 
test_regularizers.py ...
================== 3 passed in 10.78 seconds ==================================
However, I am experiencing an issue to get the shape of a tensor in the class _EigenvalueRegularizer_: the type returned for shape is different for Theano backend (Theano tensor type) and TF backend (TF TensorShape). Maybe we have to revise the files _theano_backend.py_ and _tensorflow_backend.py_
, to uniformize the output?

File: keras/regularizers.py (hunk 6 above)
Comment (oswaldoludwig) [inline keras/regularizers.py]: I included the space before the in-line comments.

**Comment by oswaldoludwig** [comment]
It seems that everything is OK with this regularizer running on Theano backend, but some error rises when applying dot() on TF backend. Someone has any tip on this? I didn't figure this out yet:

keras/regularizers.py:53: in __call__
    domin_eigenvect = K.dot(WW, o)
keras/backend/tensorflow_backend.py:214: in dot
    out = tf.matmul(x, y)
../../../miniconda/envs/test-environment/lib/python2.7/site-packages/tensorflow/python/ops/math_ops.py:951: in matmul
    name=name)
../../../miniconda/envs/test-environment/lib/python2.7/site-packages/tensorflow/python/ops/gen_math_ops.py:686: in _mat_mul
    transpose_b=transpose_b, name=name)
../../../miniconda/envs/test-environment/lib/python2.7/site-packages/tensorflow/python/ops/op_def_library.py:655: in apply_op
    op_def=op_def)
../../../miniconda/envs/test-environment/lib/python2.7/site-packages/tensorflow/python/framework/ops.py:2042: in create_op
    set_shapes_for_outputs(ret)
../../../miniconda/envs/test-environment/lib/python2.7/site-packages/tensorflow/python/framework/ops.py:1528: in set_shapes_for_outputs
    shapes = shape_func(op)
../../../miniconda/envs/test-environment/lib/python2.7/site-packages/tensorflow/python/ops/common_shapes.py:83: in matmul_shape
    b_shape = op.inputs[1].get_shape().with_rank(2)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = TensorShape([Dimension(10)]), rank = 2
    def with_rank(self, rank):
      """Returns a shape based on `self` with the given rank.
        This method promotes a completely unknown shape to one with a
        known rank.    
        Args:
          rank: An integer.
        Returns:
          A shape that is at least as specific as `self` with the given rank.
        Raises:
          ValueError: If `self` does not represent a shape with the given `rank`.
        """
      try:
        return self.merge_with(unknown_shape(ndims=rank))
      except ValueError:
>       raise ValueError("Shape %s must have rank %d" % (self, rank))
E       ValueError: Shape (10,) must have rank 2

../../../miniconda/envs/test-environment/lib/python2.7/site-packages/tensorflow/python/framework/tensor_shape.py:616: ValueError

**Comment by oswaldoludwig** [comment]
The new regularizer can be used now on both Theano and TensorFlow backends.

File: keras/regularizers.py [hunk 8]
```diff
@@ -16,6 +17,45 @@ def get_config(self):
...
+        k = self.k
+        
+        # power method for approximating the dominant eigenvector:
+        o = K.ones([dim1, 1])  # initial values for the dominant eigenvector
+        domin_eigenvect = K.dot(WW, o)
+        for n in range(power - 1):
+            domin_eigenvect = K.dot(WW, domin_eigenvect)    
+        
+        WWd = K.dot(WW, domin_eigenvect)
+        domin_eigenval = K.dot(K.transpose(WWd), domin_eigenvect) / K.dot(K.transpose(domin_eigenvect), domin_eigenvect)  # the corresponding dominant eigenvalue
```
Comment (fchollet) [inline keras/regularizers.py]: This line is too long (especially with the commend at the end)

File: keras/regularizers.py [hunk 9]
```diff
@@ -16,6 +17,45 @@ def get_config(self):
...
+        self.uses_learning_phase = True
+
+    def set_param(self, p):
+        self.p = p
+
+    def __call__(self, loss):
+        power = 9  # number of iterations of the power method
+        W = self.p
+        WW = K.dot(K.transpose(W), W)
+        dim1, dim2 = K.eval(K.shape(WW))  # number of neurons in the layer
```
Comment (fchollet) [inline keras/regularizers.py]: If this regularizer is only available for a specific format of tensor parameter, a check should verify it (also note that using `K.eval` on a shape may break if the shape cannot be computed)

File: keras/regularizers.py [hunk 10]
```diff
@@ -16,6 +17,45 @@ def get_config(self):
         return {'name': self.__class__.__name__}
 
 
+class EigenvalueRegularizer(Regularizer):
+    """This class implements the Eigenvalue Decay regularizer.
+    
+    Args:
```
Comment (fchollet) [inline keras/regularizers.py]: The dosctring format should match other docstrings found throughout the codebase.

File: keras/regularizers.py (hunk 8 above)
Comment (oswaldoludwig) [inline keras/regularizers.py]: The comment was relocated up.

File: keras/regularizers.py (hunk 9 above)
Comment (oswaldoludwig) [inline keras/regularizers.py]: It is done. The extension of this regularizer to convolutional layers (and the respective analysis) may be an interesting study, I will try to build the framework for this analysis, maybe also based on the classification margin.

File: keras/regularizers.py (hunk 10 above)
Comment (oswaldoludwig) [inline keras/regularizers.py]: I have changed the docstring format, let me know if it is acceptable now.

**Comment by fchollet** [comment]
LGTM, thanks!

**Comment by oswaldoludwig** [comment]
Thank you too!

**Pull request merged**